core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -679,7 +687,24 @@ else if (!found && result) {
                     return new ComparisonExpression(ComparisonExpression.Operator.EQUAL, toExpression(value, type), simplifiedExpressionValues.get(0));
                 }
 
-                return new InPredicate(toExpression(value, type), new InListExpression(simplifiedExpressionValues));
+                Expression simplifiedValue = toExpression(value, type);
+                if (simplifiedValue.equals(node.getValue())) {
+                    simplifiedValue = node.getValue();
+                }
+
+                Expression simplifiedValueList = new InListExpression(simplifiedExpressionValues);
+                if (simplifiedValueList.equals(node.getValueList())) {
+                    simplifiedValueList = node.getValueList();
+                }
+
+                if (simplifiedValue == node.getValue() && simplifiedValueList == node.getValueList()) {
+                    // Do not create a new instance of InPredicate expression if it would be same as original expression.
+                    // Creating a new instance of InPredicate would cause expression type cache miss, which
```

Comment:
The cache is scoped to ExpressionInterpreter. Does this change alone bring benefits, or does it require changing the cache lifecycle first?

Reply:
This fixes `io.trino.sql.planner.ExpressionInterpreter#inListCache`, but will be needed later for caching of `ExpressionInterpreter` optimizations